organizations/tests.py

```diff
@@ -39,3 +49,57 @@ def test_superuser_can_list_organizations(self):
         self.user.save()
         response = self.client.get(reverse("organizations-list"))
         self.assertEqual(response.status_code, status.HTTP_200_OK)
+
+    def test_updating_organization_recreates_user_instance_cache(self):
+        from users.models import User
+        from rest_framework.test import APIClient
+        from plio.tests import get_new_access_token
+
+        superadmin_client = APIClient()
+        superadmin_user = User.objects.create(mobile="+919988776655", is_superuser=True)
+        superadmin_access_token = get_new_access_token(
+            superadmin_user, self.application
+        )
+        superadmin_client.credentials(
+            HTTP_AUTHORIZATION="Bearer " + superadmin_access_token.token
+        )
+
+        # verify cache data doesn't exist by default
+        cache_key_name = get_cache_key(self.user)
+
+        # make a get request
+        self.client.get(reverse("users-detail", kwargs={"pk": self.user.id}))
+        # verify cache data exists as we made a GET request for user details
+        self.assertEqual(len(cache.keys(cache_key_name)), 1)
+        self.assertEqual(len(cache.get(cache_key_name)["organizations"]), 0)
+
+        # associate the current user with the organization
+        OrganizationUser.objects.create(
+            organization=self.organization_2, user=self.user, role=self.org_admin_role
+        )
+
+        # check cache after updating organization user association
+        # this time the user should have an organization associated
+        self.assertEqual(len(cache.get(cache_key_name)["organizations"]), 1)
+        self.assertEqual(
+            cache.get(cache_key_name)["organizations"][0]["name"],
+            self.organization_2.name,
+        )
+
+        # make an update to the organization name. Only plio superadmin can do it!
+        org_new_name = "Org New Name"
+        superadmin_client.patch(
+            reverse("organizations-detail", kwargs={"pk": self.organization_2.id}),
+            {"name": org_new_name},
+        )
+
+        # user cache should be deleted after organization update
+        self.assertEqual(len(cache.keys(cache_key_name)), 0)
+
+        # request user again so that we can check if the cache is updated
+        self.client.get(reverse("users-detail", kwargs={"pk": self.user.id}))
+
+        # verify cache data has now the updated value
+        self.assertEqual(
+            cache.get(cache_key_name)["organizations"][0]["name"], org_new_name
+        )
```

Comment:
one point of distinction which I feel needs to be made:
for the request above, there was a PUT request to OrganisationUser and hence, the retrieval from DB post cache invalidation happened in the same request as the request had to return the new OrganisationUser object in the response. 

However, here, for a PATCH request, once the cache is invalidated, the same request won't fetch the updated object from DB and hence, we need to make a separate GET request to fill the cache. 

This also means that the workflow diagram shouldn't be mentioning PUT/PATCH/DELETE in the types of requests for which the diagram is valid right? We should have a separate cache invalidation diagram for PATCH.

Reply:
I think you are right. There's something here that's not clearly mentioned in the diagram.

When we're making a PUT/PATCH/DELETE request for the same resource (like user in this test case), it updates the cache with these steps:
1. post_save signal deletes the cache
2. the serializer for that resource creates the cache with new values

But when we're making a PUT/PATCH/DELETE request for a related resource and not the same (like organizations for user in this case), it only deletes the user cache and not creates it again as UserSerializer didn't run in organization update/delete request.

So the diagram is partially correct. I'll see how I can fit it in the diagram.

Reply:
Yep, that would be helpful. Although I don't seem to see any changes to the image. is that intentional? to convey this just in the markdown files?

Reply:
Yep. I am a bit not sure exactly how to represent it diagrammatically. It felt like the explanation is text-heavy.